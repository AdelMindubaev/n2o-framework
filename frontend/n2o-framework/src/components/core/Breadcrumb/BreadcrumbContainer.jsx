import React from 'react';
import PropTypes from 'prop-types';
import { compose, getContext, withProps } from 'recompose';
import map from 'lodash/map';

import { getModelSelector } from '../../../selectors/models';
import propsResolver from '../../../utils/propsResolver';

/**
 * Контейнер для {@link Breadcrumb}
 * @reactProps {object} items -массив из {label, href} объектов, описывающих одну вкладку ("крошку")
 *
 * @example
 *  const items = [
 *    {
 *      href: '/test1',
 *      label: 'test1'
 *    },
 *    {
 *      href: '/test2',
 *      label: 'test2'
 *    },
 *  ]
 * <Breadcrumb  items={items}/>
 * */
function BreadcrumbContainer(props) {
  const DefaultBreadcrumb = props.defaultBreadcrumb;
  return (
    <React.Fragment>
      <DefaultBreadcrumb items={props.items} />
    </React.Fragment>
  );
}

BreadcrumbContainer.propTypes = {
  defaultBreadcrumb: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      modelLink: PropTypes.string,
      path: PropTypes.string,
    })
  ),
};

BreadcrumbContainer.defaultProps = {
  items: [],
  defaultBreadcrumb: () => null,
};

export default compose(
  getContext({
    store: PropTypes.object,
    defaultBreadcrumb: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  }),
  withProps(props => ({
    items: map(props.items, item => {
      if (item.modelLink) {
        return {
          ...item,
          label: propsResolver(
            item.label,
            getModelSelector(item.modelLink)(props.store.getState())
          ),
        };
      }

      return item;
    }),
  }))
)(BreadcrumbContainer);
