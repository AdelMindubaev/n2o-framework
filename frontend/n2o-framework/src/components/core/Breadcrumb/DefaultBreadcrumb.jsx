import React from 'react'
import { Link } from 'react-router-dom'
import map from 'lodash/map'
import cn from 'classnames'
import Breadcrumb from 'reactstrap/lib/Breadcrumb'
import BreadcrumbItem from 'reactstrap/lib/BreadcrumbItem'

function DefaultBreadcrumb({ items }) {
    const crumbs = map(items, ({ label, path }, index) => {
        const lastCrumb = index === items.length - 1

        return (
            <BreadcrumbItem key={index} active={lastCrumb}>
                {path && !lastCrumb ? (
                    <Link className="n2o-breadcrumb-link" to={path} key={index}>
                        {label}
                    </Link>
                ) : (
                    <span
                        className={cn({
                            'n2o-breadcrumb-link n2o-breadcrumb-link_active': lastCrumb,
                            'n2o-breadcrumb-link n2o-breadcrumb-link_disabled': !lastCrumb,
                        })}
                        key={index}
                    >
                        {label}
                    </span>
                )}
            </BreadcrumbItem>
        )
    })

    return <Breadcrumb>{crumbs}</Breadcrumb>
}

export default DefaultBreadcrumb
