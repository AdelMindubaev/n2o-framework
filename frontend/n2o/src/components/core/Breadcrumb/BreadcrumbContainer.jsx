import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { map, memoize } from 'lodash';

import { getModelSelector } from '../../../selectors/models';
import propsResolver from '../../../utils/propsResolver';

/**
 * Контейнер для {@link Breadcrumb}
 * @reactProps {object} items -массив из {label, href} объектов, описывающих одну вкладку ("крошку")
 *
 * @example
 *  const items = [
 *    {
 *      href: '/test1',
 *      label: 'test1'
 *    },
 *    {
 *      href: '/test2',
 *      label: 'test2'
 *    },
 *  ]
 * <Breadcrumb  items={items}/>
 * */
class BreadcrumbContainer extends React.Component {
  constructor(props) {
    super(props);
  }

  /**
   * Базовый рендер
   * @returns {*}
   */
  render() {
    const DefaultBreadcrumb = this.props.defaultBreadcrumb;
    return (
      <React.Fragment>
        <DefaultBreadcrumb items={this.props.items} />
      </React.Fragment>
    );
  }
}

BreadcrumbContainer.propTypes = {
  defaultBreadcrumb: PropTypes.node,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      modelLink: PropTypes.string,
      path: PropTypes.string
    })
  )
};

BreadcrumbContainer.defaultProps = {
  items: [],
  defaultBreadcrumb: () => null
};

const memoizedMap = memoize(map);

const mapStateToProps = (state, ownProps) => {
  return {
    items: memoizedMap(ownProps.items, item => {
      if (item.modelLink) {
        return {
          ...item,
          label: propsResolver(item.label, getModelSelector(item.modelLink.link)(state))
        };
      }

      return item;
    })
  };
};

BreadcrumbContainer = connect(mapStateToProps)(BreadcrumbContainer);
export default BreadcrumbContainer;
