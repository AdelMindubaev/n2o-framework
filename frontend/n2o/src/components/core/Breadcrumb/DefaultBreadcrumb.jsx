import React from 'react';
import { Link } from 'react-router-dom';
import { Breadcrumb, BreadcrumbItem } from 'reactstrap';

function DefaultBreadcrumb({ items }) {
  const crumbs = items.map(({ label, path }, index) => {
    return (
      <BreadcrumbItem active={index === items.length - 1}>
        {index !== items.length - 1 ? <Link to={path}>{label}</Link> : label}
      </BreadcrumbItem>
    );
  });
  return <Breadcrumb>{crumbs}</Breadcrumb>;
}

export default DefaultBreadcrumb;
