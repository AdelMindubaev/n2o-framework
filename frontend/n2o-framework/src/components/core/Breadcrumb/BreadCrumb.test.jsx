import React from 'react'
import renderer from 'react-test-renderer'
import { StaticRouter } from 'react-router'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'

import BreadcrumbContainer from './BreadcrumbContainer'
import DefaultBreadcrumb from './DefaultBreadcrumb'

const setupContainer = (propOverrides) => {
    const props = {
    // use this to assign some default props

        ...propOverrides,
    }
    const initialState = {
        models: {
            test: {
                test: 'test',
            },
        },
    }
    const mockStore = configureMockStore()
    const store = mockStore(initialState)
    const component = (
        <StaticRouter location="test" context={{}}>
            <Provider store={store}>
                <BreadcrumbContainer
                    defaultBreadcrumb={props => <DefaultBreadcrumb {...props} />}
                    {...props}
                />
            </Provider>
        </StaticRouter>
    )

    return {
        props,
        component,
    }
}

describe('<Breadcrumb />', () => {
    it('тестирует снэпшот контейнера брэдкрамба', () => {
        const { component } = setupContainer({
            items: [
                {
                    label: '`test ? test : \'dummy\'`',
                    modelLink: 'models.test',
                    path: '/redirectTest',
                },
                {
                    label: '`test == \'fake\' ? test : \'Test\'`',
                    modelLink: 'models.test',
                },
            ],
        })
        const tree = renderer.create(component).toJSON()

        expect(tree).toMatchSnapshot()
    })
})
