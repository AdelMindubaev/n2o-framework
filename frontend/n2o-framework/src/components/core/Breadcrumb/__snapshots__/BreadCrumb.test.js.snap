// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumb /> тестирует снэпшот контейнера брэдкрамба 1`] = `
<nav
  aria-label="breadcrumb"
  className=""
>
  <ol
    className="breadcrumb"
  >
    <li
      className="breadcrumb-item"
    >
      <a
        className="n2o-breadcrumb-link"
        href="/redirectTest"
        onClick={[Function]}
      >
        test
      </a>
    </li>
    <li
      aria-current="page"
      className="active breadcrumb-item"
    >
      <span
        className="n2o-breadcrumb-link n2o-breadcrumb-link_active"
      >
        Test
      </span>
    </li>
  </ol>
</nav>
`;
